src/ApacheOrcDotNet.OptimizedReaderTest.App/TradeDataSource.cs

```diff
@@ -0,0 +1,127 @@
+﻿using ApacheOrcDotNet.OptimizedReader;
+using ApacheOrcDotNet.OptimizedReader.Buffers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Reader = ApacheOrcDotNet.OptimizedReader.OrcReader;
+
+namespace ApacheOrcDotNet.OptimizedReaderTest.App
+{
+    public class TradeDataSource
+    {
+        private readonly Reader _reader;
+        private readonly string _souce;
+        private readonly string _symbol;
+
+        public TradeDataSource(Reader reader, string source, string symbol)
+        {
+            _reader = reader;
+            _souce = source;
+            _symbol = symbol;
+        }
+
+        public TimeRangeReader CreateTimeRangeReader(TimeSpan startTime, TimeSpan endTime)
+            => new TimeRangeReader(_reader, _souce, _symbol, startTime, endTime);
+    }
+
+    public class TimeRangeReader
+    {
+        private readonly Reader _reader;
+        private readonly string _souce;
+        private readonly string _symbol;
+        private readonly TimeSpan _startTime;
+        private readonly TimeSpan _endTime;
+        private readonly BaseColumnBuffer<string> _sourceColumnBuffer;
+        private readonly BaseColumnBuffer<string> _symbolColumnBuffer;
+        private readonly BaseColumnBuffer<decimal?> _timeColumnBuffer;
+        private readonly BaseColumnBuffer<decimal?> _priceColumnBuffer;
+        private readonly BaseColumnBuffer<long?> _sizeColumnBuffer;
+        private readonly Dictionary<int, List<int>> _filters = new();
+
+        public TimeRangeReader(Reader reader, string source, string symbol, TimeSpan startTime, TimeSpan endTime)
+        {
+            _reader = reader;
+            _souce = source;
+            _symbol = symbol;
+            _startTime = startTime;
+            _endTime = endTime;
+
+            var sourceColumn = _reader.GetColumn("source");
```

Comment:
We should be able to move a lot of this out to `TradeDataSource`.  Using `TimeRangeReader` should do as little work as possible.  Right now it's redoing all the work every time you request a different time range.

Reply:
This class was refactored to include this change and the ones from the other notes. The new output is as follows:

```
.
Read 27 rows of data
..
Read 14 rows of data
..........
Read 81601 rows of data
.........
Read 79387 rows of data

Read execution time: 00:00:519
Total execution time: 00:00:674
```